receiver/datadogreceiver/receiver.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2021, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package datadogreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/datadogreceiver"
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"sync"
+
+	datadogpb "github.com/DataDog/datadog-agent/pkg/trace/exportable/pb"
+	"github.com/gorilla/mux"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/obsreport"
+	"go.opentelemetry.io/collector/receiver"
+)
+
+type datadogReceiver struct {
+	config       *Config
+	params       receiver.CreateSettings
+	nextConsumer consumer.Traces
+	server       *http.Server
+	shutdownWG   sync.WaitGroup
+	tReceiver    *obsreport.Receiver
+
+	startOnce sync.Once
+	stopOnce  sync.Once
+}
+
+func newDataDogReceiver(config *Config, nextConsumer consumer.Traces, params receiver.CreateSettings) (receiver.Traces, error) {
+	if nextConsumer == nil {
+		return nil, component.ErrNilNextConsumer
+	}
+
+	instance, err := obsreport.NewReceiver(obsreport.ReceiverSettings{LongLivedCtx: false, ReceiverID: config.ReceiverID, Transport: "http", ReceiverCreateSettings: params})
+	if err != nil {
+		return nil, err
+	}
+	return &datadogReceiver{
+		params:       params,
+		config:       config,
+		nextConsumer: nextConsumer,
+		server: &http.Server{
+			ReadTimeout: config.ReadTimeout,
+			Addr:        config.HTTPServerSettings.Endpoint,
+		},
+		tReceiver: instance,
+	}, nil
+}
+
+func (ddr *datadogReceiver) Start(_ context.Context, host component.Host) error {
+	go ddr.startOnce.Do(func() {
+		defer ddr.shutdownWG.Done()
+		ddmux := mux.NewRouter()
+		ddmux.HandleFunc("/v0.3/traces", ddr.handleTraces)
+		ddmux.HandleFunc("/v0.4/traces", ddr.handleTraces)
+		ddmux.HandleFunc("/v0.5/traces", ddr.handleTraces)
+		ddr.server.Handler = ddmux
+		if err := ddr.server.ListenAndServe(); err != http.ErrServerClosed {
+			host.ReportFatalError(fmt.Errorf("error starting datadog receiver: %w", err))
+		}
+	})
+	ddr.shutdownWG.Add(1)
```

Comment:
I think an alternative way to write this is:

```suggestion
func (ddr *datadogReceiver) Start(_ context.Context, host component.Host) error {
	ddr.startOnce.Do(func() {
		ddr.shutdownWG.Add(1)

		go func() {
			defer ddr.shutdownWG.Done()
			
			ddmux := mux.NewRouter()
			ddmux.HandleFunc("/v0.3/traces", ddr.handleTraces)
			ddmux.HandleFunc("/v0.4/traces", ddr.handleTraces)
			ddmux.HandleFunc("/v0.5/traces", ddr.handleTraces)
			ddr.server.Handler = ddmux
			if err := ddr.server.ListenAndServe(); err != http.ErrServerClosed {
				host.ReportFatalError(fmt.Errorf("error starting datadog receiver: %w", err))
			}
		}()

	})
	return nil
}
```

Reply:
Ah oups, just saw this after posting https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/5836#discussion_r1088556838. I'm like a dog in a bowling alley. Forgive me!

Reply:
Hahaha, you're fine. 

I like to see it as everyone is excited to see this in.

Reply:
Adopted!